Alloy/Encoders/TextureCopyEncoder.swift

```diff
@@ -51,18 +52,57 @@ final public class TextureCopyEncoder {
     public func encode(sourceTexture: MTLTexture,
                        destinationTexture: MTLTexture,
                        using encoder: MTLComputeCommandEncoder) {
+        self.copy(region: sourceTexture.region,
+                  from: sourceTexture,
+                  to: .zero,
+                  of: destinationTexture,
+                  using: encoder)
+    }
+
+    public func copy(region sourceTexureRegion: MTLRegion,
+                     from sourceTexture: MTLTexture,
+                     to destinationTextureOrigin: MTLOrigin,
+                     of destinationTexture: MTLTexture,
+                     in commandBuffer: MTLCommandBuffer) {
+        commandBuffer.compute { encoder in
+            encoder.label = "Texture Copy"
+            self.copy(region: sourceTexureRegion,
+                      from: sourceTexture,
+                      to: destinationTextureOrigin,
+                      of: destinationTexture,
+                      using: encoder)
+        }
+    }
+
+    public func copy(region sourceTexureRegion: MTLRegion,
+                     from sourceTexture: MTLTexture,
+                     to destinationTextureOrigin: MTLOrigin,
+                     of destinationTexture: MTLTexture,
+                     using encoder: MTLComputeCommandEncoder) {
+        let readOffset = SIMD2<UInt16>(x: UInt16(sourceTexureRegion.origin.x),
+                                       y: UInt16(sourceTexureRegion.origin.y))
+        let writeOffset = SIMD2<UInt16>(x: UInt16(destinationTextureOrigin.x),
+                                        y: UInt16(destinationTextureOrigin.y))
+
         encoder.set(textures: [sourceTexture,
                                destinationTexture])
+        encoder.set(readOffset,
+                    at: 0)
+        encoder.set(writeOffset,
+                    at: 1)
+
+        let gridSize = MTLSize(width: destinationTexture.width - destinationTextureOrigin.x,
```

Comment:
this may be <= 0 and I think we should avoid launching anything

in fact, destination origin can be negative and it will increase the amount of threads. I think we should keep that in mind as well

also, what if write area exceeds read area?